extension/src/time_series/pipeline/aggregation.rs

```diff
@@ -670,4 +879,85 @@ mod tests {
+
+            let val = client.select(
+                &format!("SELECT series -> sort() -> counter_agg() -> with_bounds('[2020-01-01 UTC, 2020-02-01 UTC)') -> extrapolated_delta('prometheus') FROM ({}) s", create_series),
+                None,
+                None
+            )
+                .first()
+                .get_one::<f64>().unwrap();
+            assert_eq!(val, 67.5);
+        })
+    }
+
+    #[pg_test]
+    fn test_hyperloglog_finalizer() {
+        Spi::execute(|client| {
+            client.select("SET timezone TO 'UTC'", None, None);
+            // using the search path trick for this test b/c the operator is
+            // difficult to spot otherwise.
+            let sp = client.select("SELECT format(' %s, toolkit_experimental',current_setting('search_path'))", None, None).first().get_one::<String>().unwrap();
+            client.select(&format!("SET LOCAL search_path TO {}", sp), None, None);
+            client.select("SET timescaledb_toolkit_acknowledge_auto_drop TO 'true'", None, None);
+
+            // we use a subselect to guarantee order
+            let create_series = "SELECT timeseries(time, value) as series FROM \
+            (VALUES ('2020-01-04 UTC'::TIMESTAMPTZ, 10.0), \
+                ('2020-01-01 UTC'::TIMESTAMPTZ, 15.0), \
+                ('2020-01-03 UTC'::TIMESTAMPTZ, 20.0), \
+                ('2020-01-02 UTC'::TIMESTAMPTZ, 25.0), \
+                ('2020-01-05 UTC'::TIMESTAMPTZ, 30.0), \
+                ('2020-01-06 UTC'::TIMESTAMPTZ, 25.0), \
+                ('2020-01-07 UTC'::TIMESTAMPTZ, 15.0), \
+                ('2020-01-08 UTC'::TIMESTAMPTZ, 35.0), \
+                ('2020-01-09 UTC'::TIMESTAMPTZ, 10.0), \
+                ('2020-01-10 UTC'::TIMESTAMPTZ, 5.0)) as v(time, value)";
+
+        let val = client.select(
+            &format!("SELECT (series -> hyperloglog(100))::TEXT FROM ({}) s", create_series),
+            None,
+            None
+        )
+            .first()
+            .get_one::<String>();
+            assert_eq!(val.unwrap(), "(version:1,log:Sparse(num_compressed:7,element_type:FLOAT8,collation:None,compressed_bytes:28,precision:7,compressed:[136,188,20,7,8,30,244,43,72,69,89,2,72,255,97,27,72,83,248,27,200,110,35,5,8,37,85,12]))");
+
+            let val = client.select(
+                &format!("SELECT series -> hyperloglog(100) -> distinct_count() FROM ({}) s", create_series),
+                None,
+                None
+            )
+                .first()
+                .get_one::<i32>().unwrap();
+            assert_eq!(val, 7);
+        })
+    }
```

Comment:
If we're adding the pipeline-folding optimizations we should also add tests that they're working.